src/Streamly/Internal/Data/Ring/Foreign.hs

```diff
@@ -506,3 +509,48 @@ unsafeFoldRingNM count rh f z rb@Ring {..} =
         if ptr == rh || n == 0
             then return acc'
             else go (n - 1) acc' ptr
+
+data Tuple4' a b c d = Tuple4' !a !b !c !d deriving Show
+
+-- | @rollWindow n collector@ is a simple, incremental sliding window 
+-- fold that does not require all the intermediate elements in a computation.
+-- This maintains n elements in the window, when a new element comes it slides
+-- out the oldest element and the new element along with the old element are 
+-- supplied to the collector fold. 
+--
+-- The Maybe is for the case when initially the window is filling and 
+-- there is no old element.
+--
+{-# INLINE rollWindow #-}
+rollWindow :: forall m a b. (MonadIO m, Storable a) 
+    => Int -> Fold m (a, Maybe a) b -> Fold m a b
+rollWindow n (Fold step1 initial1 extract1)= Fold step initial extract
+
+    where
+
+    initial = do 
+        s <- initial1
+        (rb, rh) <- liftIO $ new n
+        return (Fold.Partial $ Tuple4' rb rh (0 :: Int) s)
```

Comment:
We should not just tuck in `s` in the state. We have to examine `s` and if it returns `Done` we have to return `Done` right away. If it is Partial we have to store the state inside Partial in the tuple, do not store the state boxed inside Partial/Done constructor. You delayed inspection of the result of `initial1` to the `step` function which is incorrect.

See how other such combinators are implemented in the Fold module.